rust/ql/lib/codeql/files/FileSystem.qll

```diff
@@ -33,4 +33,19 @@ class Folder = Impl::Folder;
 class File extends Container, Impl::File {
   /** Holds if this file was extracted from ordinary source code. */
   predicate fromSource() { any() }
+
+  /**
+   * Gets the number of lines containing code in this file. This value
+   * is approximate.
+   */
+  int getNumberOfLinesOfCode() {
+    result =
+      count(int line |
+        exists(Location loc |
+          loc.getFile() = this and
+          line = [loc.getStartLine(), loc.getEndLine()] and
+          not loc instanceof EmptyLocation
+        )
+      )
+  }
```

Comment:
I would try to only count lines when the thing the location is attached to was properly extracted, for which we should exclude unimplemented and missing stuff. I've opened https://github.com/github/codeql/pull/17479 to get that information in the DB. WIth that I think this could be something like
```suggestion
  int getNumberOfLinesOfCode() {
    result =
      count(int line |
        exists(Locatable e, Location loc |
          loc = e.getLocation() and
          loc.getFile() = this and
          line = [loc.getStartLine(), loc.getEndLine()] and
          not e instanceof Unextracted and
          not loc instanceof EmptyLocation
        )
      )
  }
```

Reply:
That's a good idea.  I'm keen to get these queries merged though, do you mind if I address this as follow-up?

Reply:
(I've created an issue to track this; too many open interacting PRs right now)

Reply:
I'm ok with this being a follow-up :+1: